### ISSUE TITLE
Power BI workspace not found when name contains URI reserved chars

### ISSUE THREAD

**Issue by albertospelta** [sql-bi/Bravo#500]
### Describe the problem

Copied from https://github.com/sql-bi/Bravo/issues/494#issuecomment-1277016034

```
Error 10: A connection problem arises between the server and Bravo.
The specified Power BI workspace is not found.

Technical Details:
RootActivityId: 75f75078-9984-42f6-a6b0-4e16014a7433
Trace Id: 00-10a82270feac38b0b634054e0475f894-9106091162b6ddf0-00
Version: 0.9.6 (0.9.8291.23631)
```

Got this when trying to connect to a workspace dataset that has a \ within the workspace name. Slash at end is ok, but within no chocolates. Prob a low user case but WS names are free text really!!??
I guess your not using the workspace id to pass request thru?
(Let me know if want to raise as sep issue?)

Cheers

Chris

### Bravo version

n/a

### Tabular model

Power BI Dataset

### Power BI

n/a

### Steps to reproduce the behavior

1. Create (or rename) a PBI workspace with a `\` within the workspace name
2. Publish a report in the dataset
3. Run bravo and connect to the dataset

**Comment by albertospelta** [comment]
@Chris-Burgess-ACTGOV That's a bug, thanks for reporting!

The workspace name is free text and is required in order to get the connection URL. 
`powerbi://api.powerbi.com/v1.0/[tenant name]/[workspace name]`

The unique workspace object ID can be used - appended to the workspace name - only when connecting to a workspace with the same name as another workspace (duplicated workspace names).

I have just pushed a fix for this issue, if you want to test you can update to the latest `Dev` build.

**Comment by Chris-Burgess-ACTGOV** [comment]
Install of the dev build fails ☹ when updating. 
 
<img width="330" alt="image" src="https://user-images.githubusercontent.com/60373199/195727348-c7fb0313-e7ae-40d6-9187-0a9b30e946e2.png">

System.IO.FileNotFoundException: Could not load file or assembly 'Dax.Formatter, Version=1.0.0.0, Culture=neutral, PublicKeyToken=9ab7b540eb23c0a4'. The system cannot find the file specified.
File name: 'Dax.Formatter, Version=1.0.0.0, Culture=neutral, PublicKeyToken=9ab7b540eb23c0a4'
   at Sqlbi.Bravo.Startup.ConfigureServices(IServiceCollection services)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.InvokeCore(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass9_0.<Invoke>g__Startup|0(IServiceCollection serviceCollection)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.Invoke(Object instance, IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.ConfigureServicesBuilder.<>c__DisplayClass8_0.<Build>b__0(IServiceCollection services)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.UseStartup(Type startupType, HostBuilderContext context, IServiceCollection services, Object instance)
   at Microsoft.AspNetCore.Hosting.GenericWebHostBuilder.<>c__DisplayClass13_0.<UseStartup>b__0(HostBuilderContext context, IServiceCollection services)
   at Microsoft.Extensions.Hosting.HostBuilder.CreateServiceProvider()
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at Sqlbi.Bravo.Program.CreateHost() in D:\a\1\s\src\Program.Host.cs:line 75
   at Sqlbi.Bravo.Program.Main() in D:\a\1\s\src\Program.cs:line 23

Seems to install ok if uninstall existing version and install.

But then get this err on launch
<img width="322" alt="image" src="https://user-images.githubusercontent.com/60373199/195727680-41e41a0d-b78a-4bf6-9e20-522356f354cc.png">
System.UnauthorizedAccessException: Access to the path is denied.
   at System.IO.Pipes.NamedPipeClientStream.TryConnect(Int32 timeout, CancellationToken cancellationToken)
   at System.IO.Pipes.NamedPipeClientStream.ConnectInternal(Int32 timeout, CancellationToken cancellationToken, Int32 startTime)
   at System.IO.Pipes.NamedPipeClientStream.Connect(Int32 timeout)
   at Sqlbi.Bravo.Infrastructure.AppInstance.NotifyOwner() in D:\a\1\s\src\Infrastructure\AppInstance.cs:line 62
   at Sqlbi.Bravo.Program.Main() in D:\a\1\s\src\Program.cs:line 33

**Comment by albertospelta** [comment]
The first error (`FileNotFoundException`) is due to reverting the changes applied in the internal version previously used for testing. This caused a version mismatch for the `Dax.Formatter` library and it's ok to uninstall and reinstall as you did.

Regarding the `UnauthorizedAccessException` error:
- does the error occur only the first time you run Bravo after completing the installation or every time you run the application?
- when you installd the application by right-run the setup as administrator, were you prompted to enter the username and password or just confirm (yes/no button)?
- Could you try to remove and reinstall this version? Does the error occur even if this flag (see image below) is disabled in the last step of the installation?

![image](https://user-images.githubusercontent.com/1957788/196024404-46abb01a-2ea6-470e-8d1c-ebdf8184d6b8.png)

**Comment by Chris-Burgess-ACTGOV** [comment]
### - does the error occur only the first time you run Bravo after completing the installation or every time you run the application?
Every time when running with normal account.  
Note If i launch app elevated it 'appears' to work - i can get to the PowerBI service sign in screen (for Datasets on powerbi.com), but cant then sign in (using either credential) - the auth screen takes me back to sign in without any browser auth!!). 
**Can** get to local models and 'Format DAX' works (cant test the '\' in Workspace name though, obviously)

### when you installd the application by right-run the setup as administrator, were you prompted to enter the username and password or just confirm (yes/no button)?
Username/password for elevated credentials (logged in to Windows with normal account)
### Could you try to remove and reinstall this version? Does the error occur even if this flag (see image below) is disabled in the last step of the installation?
**0.9.6-internal-20221013.1-main-1822e834e3394ad4111722d1dad5490d7a9b57d6 (0.9.8321.14222)**
Remove/Reinstall/Unflagged...run through menu shortcut - **SUCCESS**. All features working - no crash.


Interesting that its appearing to also work without explicitly referencing the port 'custom' proxy   #494 - OOB setting is 'system' #

**Comment by Chris-Burgess-ACTGOV** [comment]
Now back to the Issue 500!!

@albertospelta 
Appears the fix rolled out in the DEV build 0.9.6-internal-20221013.1-main-1822e834e3394ad4111722d1dad5490d7a9b57d6 (0.9.8321.14222) **has addressed** the embedding of 'reserved characters' in the Workspace name (not tested $ or % or other randoms)

**Comment by albertospelta** [comment]
Thanks @Chris-Burgess-ACTGOV !
I'll create a new issue for the `UnauthorizedAccessException`, it's a known bug not yet completely fixed.

> when you installd the application by right-run the setup as administrator, were you prompted to enter the username and password or just confirm (yes/no button)?
> Username/password for elevated credentials (logged in to Windows with normal account)

I still have a couple of questions on this:
- if you open the Windows Task Manager (process details page) when you get the `UnauthorizedAccessException` error, how many processes for the Bravo.exe application are there? 
- are all active Bravo.exe processes running under the same user account or with different accounts?

**Comment by Chris-Burgess-ACTGOV** [comment]
Attempting to reproduce on another machine.
Have stable release installed. Updating to DEV release.
Logged in to Windows with normal account.
Open cmd 'run as administrator'. Enter admin creds.
Run msi from cmd.
Install/overwrite works. Now at the Launch Bravo for PowerBI. (Still in the elevated 'bubble')
Successful Launch.
Attempt to log into PowerBI.com (have to tick 'Authenticate in Browser' before this will log in. [Sidebar: Should that be ticked OOB or is smaller use case MFA?]
All features work (including formatting of DAX [note no 9090 custom proxy setting required.]

Close App
Open App (no longer launched withing _A bubble). Successful.
**Cannot reproduce the Issue.**
Note have to add 909 and switch to custom proxy for formatting to work on this run.
Note the successful testing was on a W10 machine of slighly diff spec
Note second machine is:
_Edition   Windows 10 Enterprise
Version  **21H2**
Installed on           ‎22/‎04/‎2022
OS build 19044.2006
Experience            Windows Feature Experience Pack 120.2212.4180.0__

While machines that experienced errors on
_Edition	Windows 10 Enterprise
Version	**20H2**
Installed on	‎21/‎03/‎2022
OS build	19042.2006
Experience	Windows Feature Experience Pack 120.2212.4180.0_


Task Manager. All running under standard account (but remember cannot reproduce, If try to overwrite on 'error' machine get the 'remove' only option (Change/Repair greyed out) so cant review unless remove, install, stable, try upgrading to dev - launch ticked)

**Summary**
Cannot reproduce the UnauthorizedAccessException issue on the 21H2 build machine.
I will attempt the same on the 20H2 (error) machine in the next 24 hours.

**Comment by Chris-Burgess-ACTGOV** [comment]
Repeated uninstall/install (stable/dev) on the 20H2 machine. 

No UnauthorizedAccessException Error.

The workspace / , proxy:9090 and use browser to auth all working fine.

A system glitch perhaps!!!

**Comment by albertospelta** [comment]
@Chris-Burgess-ACTGOV thank you for your support in testing this error. I've just created a new issue to track this `UnauthorizedAccessException` that has been reported several times but does not seem to have been definitively resolved.

> Appears the fix rolled out in the DEV build 0.9.6-internal-20221013.1-main-1822e834e3394ad4111722d1dad5490d7a9b57d6 (0.9.8321.14222) has addressed the embedding of 'reserved characters' in the Workspace name (not tested $ or % or other randoms)

I am closing this issue since the Workspace name error has been fixed.
Thanks!
